[llvm] Add API to construct a benchmark from clang invocation.

ChrisCummins opened the pull request:
This adds a new API for constructing a benchmark from a command line invocation of a compiler. For example, given the command line:

```sh
$ gcc a.c src/b.c -DNDEBUG -lm -Oz -o foo.exe
```

The `make_benchmark_from_command_line()` method interprets this command line and rewrites it to one that emits an LLVM-IR file:

```
/path/to/compilergym/clang a.c src/b.c -DNDEG -lm -Oz -emit-llvm -c -o /path/to/benchmark.bc -Xclang disable-llvm-optmzns
```

It then loads the resulting bitcode for use as a benchmark. When the user is done playing around with the benchmark, they call a special `compiler()` method on the benchmark which re-invokes the original command line, but this time using the bitcode as input:

```
/path/to/compilergym/clang /path/to/benchmark.nc -DNDEG -lm -Oz -o foo.exe
```

Putting this all together, here is an end-to-end example using CompilerGym as a substitute for LLVM's pipeline:

```py
>>> import compiler_gym
>>> env = compiler_gym.make("llvm-v0")
>>> benchmark = env.make_benchmark_from_command_line(                                                                                                                                                                                                                       
...    "gcc a.c src/b.c -DNDEBUG -lm -Oz -o foo.exe"
... )                                                                                                                                                                                                                                                                       
>>> for _ in range(100):
...     _, _, done, _ env.step(env.action_space.sample()
...     if done:
...         env.reset()
>>> env.benchmark.compile()
# Now the file foo.exe is compiled
```
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
The idea is to make it easier to integrate CompilerGym into an existing build system, as it is normally possible to dump a verbose log of all of the compiler invocations run during a build.

ChrisCummins committed:
Import third-party gccinvocation library.

ChrisCummins committed:
Split gccinvocation tests into separate test file.

ChrisCummins committed:
[llvm] Add API to construct a benchmark from clang invocation.

This adds a new API for constructing a benchmark from a command line
invocation of a compiler. For example:

    >>> benchmark = env.make_benchmark_from_command_line(
    ...     ["gcc", "in.c", "-DNDEBUG"]
    ... )

This provides control over the build, equivalent to:

    >>> benchmark = env.make_benchmark(["in.c"], copts=["-NDEBUG"])

The idea is to make it easier to integrate CompilerGym into an
existing build system, as it is normally possible to dump a verbose
log of all of the compiler invocations run during a build.

mostafaelhoushi commented on compiler_gym/envs/llvm/benchmark_from_command_line.py:
```diff
@@ -0,0 +1,75 @@
+    def __init__(self, command_line: List[str], bitcode: bytes, timeout: int):
+        uri = (
+            f"benchmark://clang-v0/{urllib.parse.quote_plus(shlex.join(command_line))}"
+        )
```
The dataset or benchmark needs to be named `clang-v0`?

mostafaelhoushi commented on compiler_gym/envs/llvm/llvm_env.py:
```diff
@@ -581,3 +592,101 @@ def fork(self):
+        For example, the command line:
+
+            >>> benchmark = env.make_benchmark_from_command_line(
+            ...     ["gcc", "-DNDEBUG", "a.c", "b.c", "-o", "foo", "-lm"]
```
I am wondering if "gcc" works here... as down below we add `-Xclang` to the arguments

ChrisCummins replied:
Yes, this function takes an argument `replace_driver` (default=True) that swaps out whatever the first argument is for clang. So this will work:

```py
>>> env.make_benchmark_from_command_line(["gcc", "in.c"])
```

and will get translated to something like:

```
/path/to/compiler_gym/llvm/bin/clang in.c -c -emit-llvm -o - -Xclang disable-llvm-optmzns
```

but this will not:

```py
>>> env.make_benchmark_from_command_line(["gcc", "in.c"], replace_driver=False)
```

The idea with `replace_driver=False` is that it allows you to use your own compiler to produce the input bitcode, providing your compiler produces compatible bitcode.

The downside is that it won't work with command line options that aren't the same between whatever compiler the user uses and our version of clang.

ChrisCummins replied:
I'll extend the docstring to explain this better

mostafaelhoushi commented on compiler_gym/third_party/gccinvocation/gccinvocation.py:
```diff
@@ -0,0 +1,210 @@
+
+        self.executable = argv[0]
+        self.progname = os.path.basename(self.executable)
+        DRIVER_NAMES = (
+            "c89",
+            "c99",
+            "cc",
+            "gcc",
+            "c++",
+            "g++",
+            "xgcc",
+            "clang",
+            "clang++",
+        )
```
so is this library literary `gccinvokcation` or is it also a generic C compiler invocator?

ChrisCummins replied:
I've hacked on it to add support for clang by adding support for a few clang-specific flags, though I'm sure I've missed some

ChrisCummins replied:
I don't want to change the name of the third-party library, though if my edits work well then I could submit them upstream and consider renaming it to "ccinvocation" or something

mostafaelhoushi reviewed (commented):
Super Cool!
I am wondering is that environment compatible with `LlvmEnv` or `GccEnv`? I mean can we directly call the existing observation or action functions on that environment?

hughleat reviewed (approved):
LGTM.
Quick question though. How does it deal with mixed .c and .o files? E.g.:
```bash
gcc foo.c bar.o -o foo
```
My guess is that bar.o will be considered a source and that clang will barf trying to bitcode-ise it. Should those be filtered out for linking?

Commit:
Tiny code style fix.

Commit:
[llvm] Improve documentation.

Commit:
Fix cmake dependency name.

Commit:
Cross-reference issue in TODO comment.

Commit:
Fix cmake dependency name.

Commit:
[llvm] Raise an error if only object file inputs.

Commit:
[llvm] Add support for commandlines with objectfile inputs.

This enables commandlines with a combination of source and objectfile
inputs to be supported.

Commit:
Fix shlex.join() backward compatibility.

Commit:
Update and fix LLVM custom benchmark tests.

ChrisCummins commented:
> LGTM. Quick question though. How does it deal with mixed .c and .o files?

Good question! I added support so that `.o` files are excluded from the bitcode environment, but are kept for linking later.

The current implementation for this will not work if the CompilerGym backend and frontend are on different machines. To fix that, we need #325.

Cheers,
Chris
